core/src/raw/path_cache.rs

```diff
@@ -0,0 +1,144 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::raw::{get_basename, get_parent};
+use crate::*;
+use async_trait::async_trait;
+use moka::sync::Cache;
+
+/// The trait required for path cacher.
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+pub trait PathQuery {
+    /// Fetch the id for the root of the service.
+    async fn root(&self) -> Result<String>;
+    /// Query the id by parent_id and name.
+    async fn query(&self, parent_id: &str, name: &str) -> Result<Option<String>>;
+}
+
+/// PathCacher is a cache for path query.
+///
+/// OpenDAL is designed for path based storage systemds, such as S3, HDFS, etc. But there are many
+/// services that are not path based, such as OneDrive, Google Drive, etc. For these services, we
+/// lookup files based on id. The lookup of id is very expensive, so we cache the path to id mapping
+/// in PathCacher.
+///
+/// # Behavior
+///
+/// The `path` in cache is always an absolute. For example, if service root is `/root/`, then the
+/// path of file `a/b` in cache will be `/root/a/b`.
+pub struct PathCacher<Q: PathQuery> {
```

Comment:
(optional) `Cache` instead of `Cacher` would be good enough for me...

Reply:
Not interested.